### PR TITLE
Fix up partially constructed submodules.

### PR DESCRIPTION
**Pull request by glyn**
When using gosub in garden-linux-release, it seems not to update the .gitmodules file correctly and results in what are sometimes known as "semi-submodules" lying around. These cause git submodule operations to barf. This PR fixes up semi-submodules into proper submodules.

If you would like to reproduce the basic problem, clone garden-linux-release, check out the branch try-gosub-take3, revert the gosub submodule to a known old state (e.g. vito's HEAD), and then run scripts/sync-submodules.sh. This fails if run twice. Afterwards try updating submodules or calling `git submodule status` to see git submodule barf.

Anyway, here is the fix up, such as it is....

Fix submodules on exit as well as entry. This will ensure that gosub
sync no longer leave "semi-submodules" lying around.

It is still helpful to fix submodules on entry too to ensure a good
starting point for syncing. For example, if the user runs an old
version of gosub, they may end up with "semi-submodules" which need
fixing up.

Note that the submodule's git repository is not deleted, although
its path is temporarily removed from the git index.

Signed-off-by: Glyn Normington gnormington@pivotal.io

### CODE REVIEW

**Comment by glyn** [comment]
Once this PR, or equivalent, has been implemented, please unblock [#99694700](https://www.pivotaltracker.com/story/show/99694700).

**Comment by vito** [comment]
Don't really like that it prints a bunch of red text in the happy path (when it runs it at exit). Is there any way to just make it idempotent/faster (and quiet when there's nothing to do)?

**Comment by glyn** [comment]
@vito This change makes sync idempotent, much faster, and cause it to print in green when it needs to do a fix. For convenience of testing and since fixing is quite a handy function on its own, I've exposed "fix" as a subcommand.

The previous version was not idempotent because it would overwrite the URL in .gitmodules with the origin URL from the submodule's git repo. This caused problems because sometimes people need a https URL in .gitmodules and something else in the submodule's git repo.

File: main.go [hunk 1]
```diff
@@ -45,6 +45,22 @@ func main() {
 			},
 			Action: sync,
 		},
+		{
+			Name:      "fix",
+			ShortName: "s",
```
Comment (vito) [inline main.go:50]: shortname here seems borked, probably copy-paste issue. maybe `f`?

**Comment by vito** [comment]
gonna go ahead and merge, will fix the shortname myself. sorry for the delay and thanks!

File: main.go (hunk 1 above)
Comment (glyn) [inline main.go:50]: Yes, thanks for correcting this.

**Pull request merged**